datafusion/core/src/physical_plan/file_format/mod.rs

```diff
@@ -489,38 +491,93 @@ impl SchemaAdapter {
             field_mappings,
         })
     }
+
+    /// Creates a `SchemaMapping` that can be used to cast or map the columns from the file schema
+    /// to the table schema, taking into account the provided projections.
+    pub fn map_schema_with_projection(
```

Comment:
I think this method should be equivalent to

```
self.map_schema(file_schema.project(projections))
```

Reply:
I would like to suggest that `map_schema_with_projection` may not be directly implemented as `self.map_schema(file_schema.project(projections))`. In the `evolved_schema` test, some columns are processed as null arrays in `SchemaMapping.map_batch`. However, we could consider performing a projection on `table_schema` when creating `SchemaAdapter`.

Reply:
>  However, we could consider performing a projection on table_schema when creating SchemaAdapter.

Sounds good to me :+1: